Added new test sample to interpolate op in OpInfo

vfdev-5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104181

Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @ngimel

vfdev-5 committed:
Added new test sample to interpolate op in OpInfo
Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

[ghstack-poisoned]

pmeier commented on torch/testing/_creation.py:
```diff
@@ -203,6 +203,11 @@ def clamp(a: float, l: float, h: float) -> float:
                 default_high=10,
             ),
         )
+        if low + 1 == high:
+            raise ValueError(
+                f"Computed low ({low}) and high ({high}) values are almost equal. "
+                "This will generate a constant tensor which may lead to potential silent bugs"
+            )
```
I don't think we should have this error. `make_tensor` is consistent with other random tensor creation ops, i.e. treating `low` as inclusive and `high` as exclusive:

```python
low = 3
a = torch.testing.make_tensor((3, 3), low=low, high=low + 1, dtype=torch.int, device="cpu")
b = torch.randint(size=(3, 3), low=low, high=low + 1)

assert (a == b).all()
```

I agree, this might hide a bug, but I don't think we make the situation any less confusing by diverging from other well-known ops.

I'm more open to a discussion about a warning here. Since this is a testing function, that shouldn't create too much unwanted noise.

peterbell10 replied:
It also looks like the case in `sample_inputs_multi_margin_loss` is intentionally creating a constant tensor and just using `make_tensor` for symmetry with the other samples.

vfdev-5 replied:
IMO, `make_target([], low=0, high=1)` should be just `torch.tensor(0, device=device)` to be clear what the sample is.

vfdev-5 replied:
Reverted this change

vfdev-5 committed:
Update base for Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4111,8 +4111,26 @@ def shape(size, rank, with_batch_channel=True):
+        rank = 2
+        yield SampleInput(
+            make_arg(shape(400, rank)),
+            shape(270, rank, False),
+            None,
+            mode,
+            False,
```
It would be more readable if the keyword variant for these parameters is used.

vfdev-5 replied:
As discussed together with Philip, I'll add keywords everywhere in this function for readability.

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4111,8 +4111,26 @@ def shape(size, rank, with_batch_channel=True):
+            device=device,
+            dtype=dtype,
+            requires_grad=requires_grad,
+            low=0 if dtype == torch.uint8 else None,
+            high=256 if dtype == torch.uint8 else None,
```
Why are we requiring specific bounds for uint8? By default it will be `[0, 10)`:

https://github.com/pytorch/pytorch/blob/567b5e5b284c40b203fcb9394a2121e0be9b2ae9/torch/testing/_creation.py#L52

vfdev-5 replied:
typical data with uint8 has [0, 255] range and not [0, 10]. This range can effect the error value for upsampling ops, thus I'm picking more representative sample instead of current one

vfdev-5 replied:
Maybe, int8 can be extended to [-128, 128], for other dtypes it is less evident to set natural/typical ranges

pmeier replied:
I'm ok with either. Can we maybe remove the condition for `low` since `0` is the default for uint8 anyway? Plus, it would be good to add a small comment why we are not going with the default `high` value for uint8.

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4144,8 +4162,20 @@ def shape(size, rank, with_batch_channel=True):
+            device=device,
+            dtype=dtype,
+            requires_grad=requires_grad,
+            low=0 if dtype == torch.uint8 else None,
+            high=256 if dtype == torch.uint8 else None,
```
Same as above.

pmeier commented on test/inductor/test_torchinductor_opinfo.py:
```diff
@@ -481,6 +481,10 @@ def wrapper_set_seed(op, *args, **kwargs):
     "linalg.solve_triangular": {"check_gradient": False},
     "linalg.lu_factor": {"check_gradient": False},
     "linalg.lu_factor_ex": {"check_gradient": False},
+    # Temporarily skip interpolate bilinear and bicubic tests:
+    "nn.functional.interpolate.bicubic": {"assert_equal": False, "check_gradient": False},
+    "nn.functional.interpolate.bilinear": {"assert_equal": False},
+    "nn.functional.upsample_bilinear": {"assert_equal": False},
```
Q: we are skipping them, because they fail without the patches later in the stack?

vfdev-5 replied:
yes, these tests are failing and later in the stack there are improvements rather then patches

vfdev-5 replied:
Yes, exactly. In the next PRs, there is uint8 test case is added into inductor tests for interpolate and upsampling ops.

We can't ensure fully ensure consistency due to various reasons reasons, thus I was adjusting atol parameter.

vfdev-5 committed:
Update base for Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

pmeier reviewed (approved):
LGTM. Thanks for cleaning the sample inputs method Victor!

vfdev-5 committed:
Update base for Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

vfdev-5 commented:
@pytorchbot merge -r

vfdev-5 commented:
@pytorchbot merge

vfdev-5 commented:
@pytorchbot rebase

vfdev-5 committed:
Update base for Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

vfdev-5 committed:
Update base for Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

vfdev-5 commented:
@kulinseth can you please help with xfailing mps tests : 
- test_output_match_nn_functional_interpolate_bilinear_cpu_uint8
It now fails with 
```
---
Mismatched elements: 219120 / 437400 (50.1%)
Greatest absolute difference: 1 at index (0, 0, 0, 4)
Greatest relative difference: 1.0 at index (0, 0, 203, 0)

To execute this test, run the following from the base repo dir:
     python test/test_mps.py -k test_output_match_nn_functional_upsample_bilinear_cpu_uint8

---
Mismatched elements: 52 / 437400 (0.0%)
Greatest absolute difference: 2 at index (0, 0, 25, 202) (up to 1 allowed)
Greatest relative difference: 0.0714285746216774 at index (1, 1, 168, 67) (up to 0 allowed)

To execute this test, run the following from the base repo dir:
     python test/test_mps.py -k test_output_match_nn_functional_interpolate_bilinear_cpu_uint8
```

I tried to add it into xfaillist here: https://github.com/pytorch/pytorch/pull/104181/commits/9d15cda3db79ca8b747882b58ac6e1ced2919ecc but it seems to still run the test (I do not have mac arm64 to test it locally).

vfdev-5 committed:
Update base for Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented:
@kulinseth we are xfailing these for now FYI

vfdev-5 commented:
@pytorchbot rebase

vfdev-5 commented:
@pytorchbot rebase

vfdev-5 commented:
@pytorchbot rebase

lezcano commented:
@pytorchbot merge -i

kit1980 commented:
@vfdev-5 @lezcano @peterbell10 
After this PR slow / linux-focal-cuda12.1-py3-gcc9-slow-gradcheck 
 started to reliably fail with out of memory https://github.com/pytorch/pytorch/actions/runs/6387185622/job/17335931146
 
 Do you think this PR can be the actual cause?

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4293,22 +4293,53 @@ def shape(size, rank, with_batch_channel=True):
+        rank = 2
+        yield SampleInput(
+            make_arg(shape(400, rank)),
+            shape(270, rank, False),
```
Increasing the first dim from 5 to 400 probably causing the memory increase. We might be able to reland with this as a reference input instead of a sample.

peterbell10 commented:
@pytorchbot revert -m "Causing OOM in slow-gradcheck" -c nosignal

vfdev-5 committed:
Added ref input functions and reduce the sample size in ordinary sample input func

vfdev-5 committed:
Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mcarilli ptrblck leslie-fang-intel voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm ngimel

[ghstack-poisoned]

vfdev-5 committed:
Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mcarilli ptrblck leslie-fang-intel voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm ngimel

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot rebase

vfdev-5 committed:
Update on "Added new test sample to interpolate op in OpInfo"


Description:
- Added new test sample to interpolate op in OpInfo
- Fixed silent issue with zero tensor test sample for uint8 dtype

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mcarilli ptrblck leslie-fang-intel voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm ngimel

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
